### ISSUE TITLE
Why non-UI threads can't be used to safely create frozen Freezable objects

### ISSUE THREAD

**Issue by ClosetBugSlayer** [dotnet/wpf#3412]
Every single DispatcherObject takes a reference to Dispatcher.CurrentDispatcher, even objects destined to be frozen (which sets the Dispatcher property back to null). Every access of Dispatcher.CurrentDispatcher will unconditionally create a Dispatcher if one wasn't created for the current thread. Dispatcher creates an HWND and other resources that will never get freed if the Dispatcher is created on a thread pool thread, because (a) a strong reference to the Dispatcher is saved in a List<> instead of a [ThreadStatic], and (b) it receives no shutdown notification because no message pump was ever running in the first place. Therefore you have these orphaned object graphs that can even cause native UI resource exhaustion and Win32Exceptions when too many get created.

In short, if you try to load a BitmapSource from a file on a thread pool thread, you're going to leak a Dispatcher even if you freeze the object.

There doesn't appear to be semantics for constructing Freezables without a Dispatcher, but for now there doesn't need to be. There are two parts to fixing this with zero disruption to existing application code:

1. Don't allocate message pump or rendering state like MediaContext, etc until actually used. Maybe do something similar to what was done to Task in the TPL: put all the extra state into an isolated state class that at minimum is only the cost of one null reference.

2. Make each Dispatcher instance a [ThreadStatic] reference and eliminate the global list of Dispatchers. This way they can be GC'd easily on thread termination even if no shutdown notification ever happens. If Dispatchers need to be discoverable, the active ones can reveal themselves when they need to call into render services, etc, and then they should only be placed in a locked weak collection.

Optionally, add an IsDispatcherCreated() method to know if a Dispatcher was created yet for the current thread. This will allow application programmers to make more sensitive decisions about whether to take UI-related actions on the current thread.

**Comment by weltkante** [comment]
> Make each Dispatcher instance a [ThreadStatic] reference and eliminate the global list of Dispatchers. This way they can be GC'd easily on thread termination even if no shutdown notification ever happens.

This is not true, dispatcher uses win32 infrastructure like HWND which is bound to a thread and cannot be properly released from GC when the thread is not running. There may also be more complex things going on if you ever were using WPF resources on that thread, which need cross-thread synchronization to be relased.

You will be leaking those native resources in the best case, cause deadlocks, memory corruption or crashes in the worst case. Dispatcher needs proper shutdown to release native resources. WinForms had similar problems in their unit testing (deadlocks because they weren't properly shutting down their unit test threads and the GC couldn't release native resources anymore).

I agree with the intention of this issue of making dispatchers play better with the rest of the .NET infrastructure, but really, dispatchers should only ever be explicitly created, not silently. Either that or they need to be very much aware that they are not explicit and avoid creating threadbound resources (e.g. throw an exception when trying to use dispatcher infrastructure that requires thread bound resources).

WPF tried to avoid STA but it failed, it created their own version of the STA problem.

**Comment by ClosetBugSlayer** [comment]
> This is not true, dispatcher uses win32 infrastructure like HWND which is bound to a thread and cannot be properly released from GC when the thread is not running a message pump.

Did you read the post? Dispatcher creates strong roots that are not necessary when no message pump will ever be created, such as when creating frozen objects on a thread pool thread. Those strong roots are not defensively designed. The DispatcherObject base class auto-allocates the Dispatcher object even when the object won't be necessary after the call to Freeze().

Before ValueTask was created, Task had the same problem. It was overallocating heap even when not necessary. The fix was to move all the special state into a separate class.

If the proposed fixed Dispatcher does not establish strong roots because it was used for nothing but creating off-thread freezables, then it can be GC'd. Nobody's application code needs to be changed, and many closet memory leaks might go away.

> Dispatcher needs proper shutdown to release native resources.

Only when it allocates them. It does not need to allocate them in the above scenario, which happens more commonly than you think. Do ordinary contract business programmers run a DispatcherFrame when doing something on a background thread? Of course not.

**Comment by weltkante** [comment]
> Did you read the post?

Yes, but did you understand mine? I'm fully agreeing with your intention, but saying that there needs to be special attention paid to the fact that some native resources cannot be properly (de)allocated via GC on a Dispatcher thread that was implicitly created (this is already the case now).

When you make the Dispatcher collectible you need to GC the internal HWND or other native resources, which will not be possible if the thread is gone or doesn't pump. I'd prefer explicit distinction between implicit dispatchers and exceptions with early failure (on allocation, not on cleanup) when someone tries to do something on a dispatcher thread that was not explicitly enabled but still requires shutdown.

That goal should be compatible with your intend of being collectible when nothing was allocated.

The problem of GC not being able to cleanup Dispatcher native resources is not theoretic, WinForms was having a lot of trouble in unit tests with their equivalent of the WPF Dispatcher being collectible and causing deadlocks because the GC could not clean things up when the thread was gone.

Anyways, I've not really much to discuss on this, I just wanted to mention it as a data point to consider when handling your issue/suggestion.

**Comment by ClosetBugSlayer** [comment]
> there needs to be special attention paid to the fact that some native resources cannot be properly (de)allocated via GC on a Dispatcher thread that was implicitly created (this is already the case now).

Fixing that is the whole point of this bug report. Implicit creation should not allocate those native resources.

>The problem of GC not being able to cleanup Dispatcher native resources is not theoretic, WinForms was having a lot of trouble in unit tests with their equivalent of the WPF Dispatcher being collectible and causing deadlocks because the GC could not clean things up when the thread was gone.

Every multithreaded application has to contend with that class of bug regardless of framework. Creating and destroying threads is a delicate process when all your objects have thread affinity.
